# **Team 17B - Kompi 17**
---
## **1 7 B (Satu 7an Bersama)**

## **Library**

In [30]:
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import CountVectorizer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import nltk
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
import pandas as pd
import numpy as np

In [31]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None) 

In [32]:
# Install NLTK
# nltk.download('punkt')
# nltk.download('stopwords')

## **Data Training**

**LINK:**
* [Health](https://www.bbc.com/news/health-47543985)
* [Sports](https://www.bbc.com/sport/football/articles/cqvn39z4yddo)
* [Finance](https://www.cnbc.com/2024/04/25/deutsche-bank-first-quarter-2024-earnings.html)


In [33]:
# Siapkan teks/artikel untuk setiap topik
health_text = "An international study suggests the pill lowers cholesterol in people who continue to have high levels despite taking other drugs such statins. And scientists suggest the new therapy may also work as an alternative for people who are unable to take statins because of side-effects. The research is published in the New England Journal of Medicine. Researchers say they have asked UK and US drug regulators to consider whether to approve the pill.Cardiovascular disease kills about 150,000 people in the UK each year. Bad cholesterol is one of the main reasons - it leads to blood vessels furring up and becoming easy to block. Blockages can be fatal - starving the heart or brain of oxygen and causing heart attacks and strokes. Cutting saturated fat and having a healthy diet, along with regular exercise, can help lower bad cholesterol. But this doesn't work for everyone. And, for some, genetic conditions - rather than lifestyle - increase their levels. Millions of people worldwide are prescribed drugs, most commonly statins, to reduce the amount of bad cholesterol in the blood. But reported side-effects and how often these drugs are prescribed has attracted controversy. The new drug works by blocking a key enzyme in the body, used to make cholesterol. Prof Sir Nilesh Samani, of the British Heart Foundation charity, says: On the whole, statins do a great job of lowering cholesterol. However, this new drug could provide real benefit for the few people who can't take them or require additional treatments to get it to the right level. The research suggests that it has the potential to reduce risk of heart attacks and strokes without major side-effects."

sports_text = "If the Gunners shared the nagging doubts that resurfaced about their character – “bottle” to use the vernacular – after a short, recent slump, they were banished by a virtuoso performance that left Chelsea humiliated at an exuberant Emirates Stadium. Arsenal’s 5-0 win put them three points ahead of Liverpool and four ahead of Manchester City, not to mention delivering a plus-56 goal difference that pretty much provides the buffer of an extra point. They have played one more game than Liverpool and two more than City but scoreboard pressure counts at this point and Arsenal are ahead on that measure. Mikel Arteta’s side wobbled in the 2-0 loss to Aston Villa in their last home league game, the pressure, tension and anticipation of going into that match having just seen Liverpool lose against Crystal Palace at Anfield earlier seemingly overcoming them. And the subsequent timid Champions League quarter-final exit to Bayern Munich raised the spectre that Arsenal were about to “choke” in the manner that agonisingly consumed them at the final fences of last season’s Premier League race. The response, a battling win at Wolverhampton Wanderers on Saturday followed by this champagne show in front of their own celebrating fans, should put to bed any suggestions that it could be a lack of character that will undermine Arsenal in the run-in. Games like this, at this stage of the season, are so often tight and settled on fine margins and details. Not here as Arsenal ran riot in the face of Chelsea’s non-existent challenge. Their performance ticked every box Arteta would have wanted before the game and then some, apart from maybe missing chances that might have put their goal difference even further out of sight."

finance_text = "Deutsche Bank shares popped to a more than six-year high on Thursday, after the German lender reported a 10% rise in first-quarter profit, beating expectations amid an ongoing recovery in its investment banking unit. Shares of Deutsche Bank provisionally ended the trading session up 8.2%. After declining in the morning, the stock price reversed course to notch its highest intraday level since December 2017, according to LSEG data. Net profit attributable to shareholders was 1.275 billion euros ($1.365 billion) for the period, ahead of an aggregate analyst forecast of 1.23 billion euros for the period, according to LSEG data. Deutsche Bank said this was its highest first-quarter profit since 2013. It also marks the bank’s 15th straight quarterly profit. Group revenue rose 1% year-on-year to 7.8 billion euros, which the bank attributed to growth in commissions and fee income, along with strength in fixed income and currencies. The revenue print also came in ahead of an analyst forecast of 7.73 billion euros, according to LSEG. Revenues at its investment bank increased 13% to 3 billion euros, following a 9% slump through full-year 2023 which had dragged down overall profit. The performance restores the division as Deutsche Bank’s highest-earning unit on growth in financing and credit trading revenue. Germany’s biggest lender reported net profit of 1.3 billion euros in the prior quarter and of 1.16 billion euros in the first quarter last year. In 2023, the bank announced it would cut 3,500 jobs over the coming years, as it targets 2.5 billion euros in operational efficiencies to boost profitability and increase shareholder returns. In a research note Thursday, analysts at Keefe, Bruyette & Woods called the group results reasonable but nothing special, highlighting strong investment bank figures but underperformance in its corporate bank and asset management divisions."

## **Data Testing**

In [34]:
new_text = "Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres."

## **Stop Words**

In [35]:
# Preprocessing teks yang diberikan
new_text_tokens = word_tokenize(new_text.lower())
stop_words_list = list(stopwords.words('english'))
new_text_tokens = [word for word in new_text_tokens if word.isalnum() and word not in stop_words_list]

`Stop word` digunakan untuk menghilangkan kata-kata yang tidak memiliki makna dalam suatu kalimat seperti kata hubung, kata ganti, dan lain-lain. Hal ini dilakukan untuk mengurangi noise pada data.

## **Count Vectorizer**

In [36]:
# Inisialisasi CountVectorizer dengan daftar stop words
vectorizer = CountVectorizer(stop_words=stop_words_list)

# Hitung vektor Bag-of-Words untuk setiap topik
X = vectorizer.fit_transform([health_text, sports_text, finance_text])
new_text_vector = vectorizer.transform([' '.join(new_text_tokens)])

`Count Vectorizer` digunakan untuk mengubah data teks menjadi vektor. `Count Vectorizer` menghitung frekuensi kemunculan kata pada data teks. Hal ini berguna untuk mengubah data teks menjadi data numerik yang dapat digunakan pada model `NLP`.

## **Unique Words**

### **UW in Every Topics**

In [37]:
# Feature names
feature_names = vectorizer.get_feature_names_out()

# Mendapatkan indeks fitur untuk setiap topik
health_indices = X[0].indices
sports_indices = X[1].indices
finance_indices = X[2].indices

# Mengambil kata-kata unik yang muncul dalam setiap topik
health_words = [feature_names[idx] for idx in health_indices]
sports_words = [feature_names[idx] for idx in sports_indices]
finance_words = [feature_names[idx] for idx in finance_indices]

# Hitung jumlah kemunculan setiap kata dalam setiap topik
total_health_words = len(health_words)
total_sports_words = len(sports_words)
total_finance_words = len(finance_words)

# Menampilkan kata-kata unik dalam setiap topik
print(f"Kata-kata unik dalam topik Health - {total_health_words} kata:")
print(health_words)

print(f"\nKata-kata dalam topik Sports - {total_sports_words} kata:")
print(sports_words)

print(f"\nKata-kata dalam topik Finance - {total_finance_words} kata:")
print(finance_words)

Kata-kata unik dalam topik Health - 127 kata:
['international', 'study', 'suggests', 'pill', 'lowers', 'cholesterol', 'people', 'continue', 'high', 'levels', 'despite', 'taking', 'drugs', 'statins', 'scientists', 'suggest', 'new', 'therapy', 'may', 'also', 'work', 'alternative', 'unable', 'take', 'side', 'effects', 'research', 'published', 'england', 'journal', 'medicine', 'researchers', 'say', 'asked', 'uk', 'us', 'drug', 'regulators', 'consider', 'whether', 'approve', 'cardiovascular', 'disease', 'kills', '150', '000', 'year', 'bad', 'one', 'main', 'reasons', 'leads', 'blood', 'vessels', 'furring', 'becoming', 'easy', 'block', 'blockages', 'fatal', 'starving', 'heart', 'brain', 'oxygen', 'causing', 'attacks', 'strokes', 'cutting', 'saturated', 'fat', 'healthy', 'diet', 'along', 'regular', 'exercise', 'help', 'lower', 'everyone', 'genetic', 'conditions', 'rather', 'lifestyle', 'increase', 'millions', 'worldwide', 'prescribed', 'commonly', 'reduce', 'amount', 'reported', 'often', 'attr

Kata diatas adalah kata yang unik pada setiap topik. Kata-kata ini digunakan untuk mengetahui kata-kata yang paling sering muncul pada setiap topik.

### **Bag of Words**

In [38]:
# Tampilkan Bag-of-Words untuk teks yang diberikan
print("Bag-of-Words untuk teks yang diberikan:")
print(feature_names)

Bag-of-Words untuk teks yang diberikan:
['000' '10' '13' '150' '15th' '16' '2013' '2017' '2023' '23' '275' '365'
 '500' '56' '73' 'according' 'additional' 'aggregate' 'agonisingly'
 'ahead' 'along' 'also' 'alternative' 'amid' 'amount' 'analyst' 'analysts'
 'anfield' 'announced' 'anticipation' 'apart' 'approve' 'arsenal' 'arteta'
 'asked' 'asset' 'aston' 'attacks' 'attracted' 'attributable' 'attributed'
 'bad' 'banished' 'bank' 'banking' 'battling' 'bayern' 'beating'
 'becoming' 'bed' 'benefit' 'biggest' 'billion' 'block' 'blockages'
 'blocking' 'blood' 'body' 'boost' 'bottle' 'box' 'brain' 'british'
 'bruyette' 'buffer' 'called' 'came' 'cardiovascular' 'causing'
 'celebrating' 'challenge' 'champagne' 'champions' 'chances' 'character'
 'charity' 'chelsea' 'choke' 'cholesterol' 'city' 'coming' 'commissions'
 'commonly' 'conditions' 'consider' 'consumed' 'continue' 'controversy'
 'corporate' 'could' 'counts' 'course' 'credit' 'crystal' 'currencies'
 'cut' 'cutting' 'data' 'december' 'decl

`Bag of Words` berisi gabungan dari kata-kata yang unik pada setiap topik. `Bag of Words` digunakan untuk mengetahui kata-kata yang paling sering muncul pada setiap topik. `Bag of Words` masih berhubungan dengan `Count Vectorizer` keduanya digunakan untuk mengubah data teks menjadi data numerik.

## **Vector Bag-of-Words**

### **Vectors in Every Topics**

In [39]:
# Buat DataFrame untuk vektor Bag-of-Words masing-masing 'Topik'
df_combined = pd.DataFrame(X.toarray().T, columns=['Health', 'Sports', 'Finance'], index=feature_names)

# Tambahkan kolom 'Target' untuk menghitung teks baru
df_combined['Target'] = new_text_vector.toarray().T

In [40]:
# Tampilkan DataFrame untuk setiap topik
print("DataFrame vector Bag-of-Words:")
print(df_combined)

DataFrame vector Bag-of-Words:
                  Health  Sports  Finance  Target
000                    1       0        0       0
10                     0       0        1       0
13                     0       0        1       0
150                    1       0        0       0
15th                   0       0        1       0
16                     0       0        1       0
2013                   0       0        1       0
2017                   0       0        1       0
2023                   0       0        2       0
23                     0       0        1       0
275                    0       0        1       0
365                    0       0        1       0
500                    0       0        1       0
56                     0       1        0       0
73                     0       0        1       0
according              0       0        3       0
additional             1       0        0       0
aggregate              0       0        1       0
agonisingly        

### **Total Vector Bag-of-Words**

In [41]:
# Jumlah total vektor Bag-of-Words untuk setiap topik
total_vectors_per_topic = X.sum(axis=1)

# Menyusun DataFrame
df_total_vectors = pd.DataFrame(total_vectors_per_topic, columns=['Total Vectors'], index=['Kesehatan', 'Olahraga', 'Keuangan'])

df_total_vectors.loc['Target'] = new_text_vector.sum()

df_total_vectors

Total Vectors
Kesehatan            167
Olahraga             162
Keuangan             202
Target                10

## **Modeling**

### **Cosinus Similarity**

`Cosinuns Similarity` digunakan untuk mengukur seberapa mirip dua text berdasarkan sudut antara dua `vektor`. `Cosinus Similarity` mengukur kemiripan antara dua text dengan menghitung cosinus dari sudut antara dua `vektor`. Semakin tinggi nilai similarity kosinus, semakin besar kedua `vektor` yang dibandingkan memiliki arah yang serupa dalam ruang `vektor`. Ini menunjukkan bahwa kedua `vektor` memiliki pola atau fitur yang serupa dalam representasi `vektor` mereka. Jadi, semakin tinggi nilai `similarity kosinus`, semakin besar kesamaan antara dua text yang dibandingkan dalam analisis yang melibatkan representasi `vektor`.

#### Manually

In [42]:
# Fungsi untuk menghitung kemiripan kosinus antara dua vektor
def cosine_similarity_manual(vector1, vector2):
    dot_product = np.dot(vector1, vector2)
    norm_vector1 = np.linalg.norm(vector1)
    norm_vector2 = np.linalg.norm(vector2)
    similarity = dot_product / (norm_vector1 * norm_vector2)
    return similarity

In [43]:
# Hitung kemiripan kosinus antara vektor teks baru dan vektor teks pelatihan
similarity_scoresM = []
for vector in X:
    similarity = cosine_similarity_manual(new_text_vector.toarray()[0], vector.toarray()[0])
    similarity_scoresM.append(similarity)

predicted_topic_index_cosineM = np.argmax(similarity_scoresM)

Fungsi `cosine_similarity_manual` menghitung `similarity cosine` antara dua vektor dengan rumus berikut:

* Pertama, dot product antara kedua vektor dihitung. Selanjutnya, norm dari masing-masing vektor dihitung. 
* Terakhir, similarity cosine dihitung dengan membagi dot product oleh hasil perkalian norm vektor. 
* Dalam loop for, fungsi ini digunakan untuk menghitung similarity antara vektor teks baru dan setiap vektor dalam kumpulan vektor X. 
* Hasil similarity disimpan dalam list, dan indeks dengan nilai similarity tertinggi digunakan untuk memprediksi topik yang paling sesuai.

#### Library

In [44]:
# Hitung kemiripan kosinus antara vektor teks baru dan vektor teks pelatihan
similarity_scoresL = cosine_similarity(new_text_vector, X)
predicted_topic_index_cosineL = similarity_scoresL.argmax()

Fungsi ini langsunng menggunakan library `cosine_similarity` dari `sklearn.metrics.pairwise` untuk menghitung `similarity cosine` antara dua vektor.

### **Experiment**

#### Multinomial Naive Bayes

MNB atau Multinomial Naive Bayes adalah salah satu bentuk dari algoritma Naive Bayes yang diadaptasi khusus untuk data yang terstruktur dalam format multiclass, seperti teks yang direpresentasikan dalam bentuk vektor. MNB mengasumsikan bahwa setiap fitur (kata) dalam data teks adalah independen, yang berarti bahwa kemunculan satu kata tidak memengaruhi kemunculan kata lainnya. MNB menghitung probabilitas kemunculan setiap kata dalam setiap kelas, dan kemudian mengalikan probabilitas tersebut untuk mendapatkan probabilitas dari seluruh data teks. MNB memilih kelas dengan probabilitas tertinggi sebagai prediksi.

* **Keunggulan**: Cepat, sederhana, dan efisien dalam memproses data teks. Toleran terhadap fitur yang jarang atau tidak ada pada dataset pelatihan.
* **Keterbatasan**: MNB menganggap setiap fitur (kata) sebagai independen, yang mungkin tidak sesuai dengan keadaan sebenarnya dalam data teks di mana konteks penting. MNB tidak dapat menangani hubungan antara kata-kata dalam data teks.

In [45]:
# Lakukan klasifikasi menggunakan Naive Bayes
classifier_nb = MultinomialNB()
classifier_nb.fit(X, [0, 1, 2])
predicted_topic_index_nb = classifier_nb.predict(new_text_vector)
prob_nb = classifier_nb.predict_proba(new_text_vector)[0]

#### Logistic Regression

Regresi Logistik adalah model statistik yang digunakan untuk klasifikasi. Ini menghitung probabilitas bahwa suatu instance termasuk dalam kelas tertentu berdasarkan fitur-fiturnya. Regresi Logistik menggunakan fungsi logistik untuk menghitung probabilitas, dan kemudian memprediksi kelas dengan probabilitas tertinggi. Regresi Logistik adalah model linier, yang berarti bahwa ia mencari hubungan linier antara fitur dan target. Regresi Logistik adalah model yang sederhana dan efisien, dan sering digunakan sebagai model dasar untuk klasifikasi.

* **Keunggulan**: Mudah diinterpretasikan, cocok untuk klasifikasi biner dan multiclass, dan relatif tahan terhadap overfitting dengan parameter yang sesuai.
* **Keterbatasan**: Kurang efisien dalam menangani dataset dengan dimensi yang sangat tinggi atau ketika hubungan antara fitur dan target tidak linear.

In [46]:
# Inisialisasi klasifikasi Logistic Regression
classifier_lr = LogisticRegression()
classifier_lr.fit(X, [0, 1, 2])
predicted_topic_index_lr = classifier_lr.predict(new_text_vector)
prob_lr = classifier_lr.predict_proba(new_text_vector)[0]

#### Random Forest

Random Forest adalah model ensemble yang terdiri dari sejumlah pohon keputusan. Setiap pohon keputusan dalam Random Forest dibangun secara independen menggunakan sub-klas dataset yang diambil secara acak. Saat melakukan prediksi, setiap pohon memberikan prediksi kelas, dan hasil akhirnya diambil melalui mayoritas suara (untuk klasifikasi) dari semua pohon.

* **Keunggulan**: Random Forest efektif dalam menangani dataset dengan fitur yang banyak dan kompleks, toleran terhadap overfitting karena penggunaan ensambel, dan relatif mudah untuk diimplementasikan. Selain itu, Random Forest dapat memberikan perkiraan probabilitas untuk prediksi kelas.
* **Keterbatasan**: Meskipun Random Forest cenderung menghasilkan model yang lebih baik daripada pohon keputusan tunggal, kompleksitas modelnya membuatnya lebih sulit untuk diinterpretasi. Selain itu, karena jumlah pohon yang digunakan dapat mempengaruhi kinerja dan waktu pelatihan, ada juga overhead komputasi yang harus dipertimbangkan.

In [57]:
classifier_rf = RandomForestClassifier(n_estimators=100, random_state=42)
classifier_rf.fit(X, [0, 1, 2])
predicted_topic_index_rf = classifier_rf.predict(new_text_vector)
prob_rf= classifier_rf.predict_proba(new_text_vector)[0]


## **Classification Results**

In [58]:
topik = ['Kesehatan', 'Olahraga', 'Keuangan']

### **Cosine Similarity**

In [59]:
# Buat DataFrame dengan kedua kolom
df_similarity_scoresM = pd.DataFrame(similarity_scoresM, columns=['Manual'], index=['Kesehatan', 'Olahraga', 'Keuangan'])
df_similarity_scoresL = pd.DataFrame(similarity_scoresL[0], columns=['Library'], index=['Kesehatan', 'Olahraga', 'Keuangan'])
df_similarity_scores = pd.concat([df_similarity_scoresM, df_similarity_scoresL], axis=1)

df_similarity_scores

Manual   Library
Kesehatan  0.016476  0.016476
Olahraga   0.224544  0.224544
Keuangan   0.048536  0.048536

In [60]:
# Output hasil klasifikasi
print("Klasifikasi menggunakan Kesamaan Kosinus Hitung Manual:", topik[predicted_topic_index_cosineM], ", dengan nilai kemiripan kosinus maksimum:", max(similarity_scoresM))
print("Klasifikasi menggunakan Kesamaan Kosinus dari Scikit-learn:", topik[predicted_topic_index_cosineL], ", dengan nilai kemiripan kosinus maksimum:", similarity_scoresL.max())

Klasifikasi menggunakan Kesamaan Kosinus Hitung Manual: Olahraga , dengan nilai kemiripan kosinus maksimum: 0.22454435656953595
Klasifikasi menggunakan Kesamaan Kosinus dari Scikit-learn: Olahraga , dengan nilai kemiripan kosinus maksimum: 0.22454435656953597


Hasil Kesamaan Kosinus new_text dalam setiap topic memiliki nilai yang berbeda-beda. Nilai similarity cosine tertinggi menunjukkan bahwa new_text memiliki kemiripan yang tinggi dengan topik tersebut. Topic Olahraga memiliki nilai similarity cosine sebesar `0.224544`, Topic Kesehatan memiliki nilai similarity cosine sebesar `0.0016476`, dan Topic Keuangan memiliki nilai similarity cosine sebesar `0.048536`. Topic Olahraga memiliki nilai similarity cosine tertinggi, sehingga new_text diprediksi sebagai topik Olahraga.

Dapat dilihat pula bahwa tidak ada perbedaan sedikitpun antara hasil `cosine similarity` yang dihitung secara manual dan menggunakan library. Hal ini menunjukkan bahwa kedua metode menghasilkan hasil yang serupa dalam mengukur kemiripan antara dua vektor. Namun, penggunaan library lebih efisien dan mudah digunakan daripada perhitungan manual.

### **Experiment**

#### Multinomial Naive Bayes

In [61]:
# Buat DataFrame untuk probabilitas Naive Bayes
df_predict_nb = pd.DataFrame(prob_nb, columns=['Probabilitas'], index=['Kesehatan', 'Olahraga', 'Keuangan'])
df_predict_nb

Probabilitas
Kesehatan      0.002092
Olahraga       0.991179
Keuangan       0.006729

In [62]:
print("Klasifikasi menggunakan Naive Bayes:", topik[predicted_topic_index_nb[0]], ", dengan probabilitas:", prob_nb[predicted_topic_index_nb[0]])

Klasifikasi menggunakan Naive Bayes: Olahraga , dengan probabilitas: 0.9911794186911391


Multinomial Naive Bayes memberikan hasil yang baik dalam klasifikasi topik berdasarkan data teks. Dalam eksperimen ini, Multinomial Naive Bayes memberikan probabilitas prediksi sebesar `0.991179` untuk topik Olahraga, `0.002092` untuk topik Kesehatan, dan `0.006729` untuk topik Keuangan. Prediksi topik Olahraga adalah yang tertinggi, sehingga new_text diprediksi sebagai topik Olahraga oleh model Multinomial Naive Bayes.

#### Logistic Regression

In [63]:
# Buat DataFrame untuk probabilitas Logistic Regression
df_predict_lr = pd.DataFrame(prob_lr, columns=['Probabilitas'], index=['Kesehatan', 'Olahraga', 'Keuangan'])
df_predict_lr

Probabilitas
Kesehatan      0.275674
Olahraga       0.628530
Keuangan       0.095796

In [64]:
print("Klasifikasi menggunakan Logistic Regression:", topik[predicted_topic_index_lr[0]], ", dengan probabilitas:", prob_lr[predicted_topic_index_lr[0]])

Klasifikasi menggunakan Logistic Regression: Olahraga , dengan probabilitas: 0.628529996048402


Logistic Regression memberikan hasil yang cukup dalam klasifikasi topik berdasarkan data teks. Dalam eksperimen ini, Logistic Regression memberikan probabilitas prediksi sebesar `0.628530` untuk topik Olahraga, `0.275674` untuk topik Kesehatan, dan `0.095796` untuk topik Keuangan. Prediksi topik Olahraga adalah yang tertinggi, sehingga new_text diprediksi sebagai topik Olahraga oleh model Logistic Regression. Probabilitas nya jauh lebih rendah dibandingkan dengan Multinomial Naive Bayes. Hal ini menunjukan performa Multinomial Naive Bayes jauh lebih baik dibandingkan dengan Logistic Regression.

#### Random Forest

In [65]:
# Buat DataFrame untuk probabilitas Support Vector Machine
df_predict_knn = pd.DataFrame(prob_rf, columns=['Probabilitas'], index=['Kesehatan', 'Olahraga', 'Keuangan'])
df_predict_knn

Skor Keputusan
Kesehatan            0.44
Olahraga             0.26
Keuangan             0.30

In [66]:
print("Klasifikasi menggunakan KNN:", topik[predicted_topic_index_rf[0]], ", dengan nilai skor keputusan:", prob_rf[predicted_topic_index_rf[0]])

Klasifikasi menggunakan KNN: Kesehatan , dengan nilai skor keputusan: 0.44


Untuk model Random Forest memberikan hasil yang kurang baik dalam klasifikasi topik berdasarkan data teks. Dalam eksperimen ini, Random Forest memberikan probabilitas prediksi sebesar `0.44` untuk topik Olahraga, `0.26` untuk topik Kesehatan, dan `0.30` untuk topik Keuangan. Prediksi topik Kesehatan adalah yang tertinggi, sehingga new_text diprediksi sebagai topik Kesehatan oleh model Random Forest. Hasil klasifikasi yang salah ini menunjukkan bahwa Random Forest tidak cocok untuk dataset ini. Hal ini kemungkinan disebabkan oleh sedikitnya dataset yang digunakan, sehingga Random Forest tidak dapat mempelajari pola yang tepat dalam data.

## **Kesimpulan**

Dari hasil eksperimen, dapat disimpulkan bahwa Multinomial Naive Bayes adalah model yang paling baik dalam klasifikasi topik berdasarkan data teks. Multinomial Naive Bayes memberikan probabilitas prediksi tertinggi untuk topik Olahraga, yang sesuai dengan hasil similarity cosine. Logistic Regression memberikan hasil yang cukup baik, namun probabilitas prediksi nya jauh lebih rendah dibandingkan dengan Multinomial Naive Bayes. Random Forest memberikan hasil yang kurang baik, dengan prediksi yang salah. Hal ini menunjukkan bahwa Multinomial Naive Bayes adalah model yang paling cocok untuk dataset ini, dan dapat digunakan untuk klasifikasi topik berdasarkan data teks.